### PR TITLE
Request Management Improvements (forked from #2966)

### PR DESCRIPTION
**Pull request by danghvu**

### CODE REVIEW

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
bot:retest

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hajimefu** [comment]
Does this supersede #2966?

**Comment by pavanbalaji** [comment]
Yes, I'll close #2966.

**Comment by pavanbalaji** [comment]
Cleaned up the patches a bit and rebased on to `mpich/master`.  Testing again.

test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/request/wait.c [hunk 1]
```diff
@@ -92,10 +92,15 @@ int MPIR_Wait_impl(MPI_Request * request, MPI_Status * status)
         MPID_Progress_end(&progress_state);
     }
 
-    mpi_errno = MPIR_Request_complete(request, request_ptr, status, &active_flag);
+    mpi_errno = MPIR_Request_completion_processing(request_ptr, status, &active_flag);
     if (mpi_errno)
         MPIR_ERR_POP(mpi_errno);
 
+    if (!MPIR_Request_is_persistent(request_ptr)) {
```
Comment (danghvu) [inline src/mpi/request/wait.c]: We cannot check error first, since in the case of error it will return without freeing the request and so leaking it.

Comment (pavanbalaji) [reply src/mpi/request/wait.c]: Oh, you are right.  My bad.  Sorry for breaking that.

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
Any other comment on this PR ? Could someone approve and/or merge this ?

**Comment by pavanbalaji** [comment]
I'll add the necessary signoffs, co-authorships, and merge.

**Pull request merged**